linera-core/src/chain_worker/cache.rs

```diff
@@ -0,0 +1,259 @@
+use linera_chain::{data_types::ExecutedBlock, types::Hashed};
+use linera_storage::Storage;
+use lru::LruCache;
+use tokio::sync::{mpsc, OwnedSemaphorePermit, Semaphore};
+use tracing::Instrument as _;
+
+use crate::{
+    chain_worker::{ChainWorkerActor, ChainWorkerConfig, ChainWorkerRequest, DeliveryNotifier},
+    join_set_ext::{JoinSet, JoinSetExt},
+    value_cache::ValueCache,
+    worker::WorkerError,
+};
+
+/// A cache of running [`ChainWorkerActor`]s.
+#[derive(Clone)]
+pub struct ChainWorkers<StorageClient>
+where
+    StorageClient: Storage,
+{
+    /// One-shot channels to notify callers when messages of a particular chain have been
+    /// delivered.
+    delivery_notifiers: Arc<Mutex<DeliveryNotifiers>>,
+    /// The set of spawned [`ChainWorkerActor`] tasks.
+    tasks: Arc<Mutex<JoinSet>>,
+    /// The cache of running [`ChainWorkerActor`]s.
+    cache: Arc<Mutex<LruCache<ChainId, ChainRequestSender<StorageClient>>>>,
+    /// Active endpoint permits.
+    active_endpoints: Arc<Semaphore>,
+}
+
+/// The map of [`DeliveryNotifier`]s for each chain.
+pub(crate) type DeliveryNotifiers = HashMap<ChainId, DeliveryNotifier>;
+
+impl<StorageClient> ChainWorkers<StorageClient>
+where
+    StorageClient: Storage,
+{
+    /// Creates a new [`ChainWorkers`] that stores at most `limit` workers.
+    pub fn new(limit: NonZeroUsize) -> Self {
+        ChainWorkers {
+            delivery_notifiers: Arc::default(),
+            tasks: Arc::default(),
+            cache: Arc::new(Mutex::new(LruCache::new(limit))),
+            active_endpoints: Arc::new(Semaphore::new(limit.get())),
+        }
+    }
+
+    /// Obtains a [`ChainActorEndpoint`] to a [`ChainWorkerActor`] for a specific [`ChainId`], or
+    /// creates a new one if it doesn't yet exist.
+    pub async fn get_endpoint(
+        &self,
+        chain_id: ChainId,
+    ) -> Result<ChainActorEndpoint<StorageClient>, NewChainActorEndpoint<StorageClient>> {
+        let permit = self
+            .active_endpoints
+            .clone()
+            .acquire_owned()
+            .await
+            .expect("`active_endpoints` semaphore should never be closed");
+
+        self.try_get_existing_endpoint(chain_id, permit).map_err(
+            |MissingEndpointError {
+                 cache_is_full,
+                 permit,
+             }| {
+                if cache_is_full {
+                    self.stop_one();
+                }
+
+                self.create_new_endpoint(chain_id, permit)
+            },
+        )
+    }
+
+    /// Attempts to get a [`ChainActorEndpoint`] for a chain worker actor that's already running.
+    fn try_get_existing_endpoint(
+        &self,
+        chain_id: ChainId,
+        permit: OwnedSemaphorePermit,
+    ) -> Result<ChainActorEndpoint<StorageClient>, MissingEndpointError> {
+        let mut cache = self.cache.lock().unwrap();
+
+        if let Some(sender) = cache.get(&chain_id) {
+            Ok(ChainActorEndpoint::new(sender.clone(), permit))
```

Comment:
(See my previous comment.) Would it be better to clone the existing permit here?